src/services/accounts/AccountsAssetsService.ts

```diff
@@ -0,0 +1,141 @@
+	 * given `AccountId`, and send back all the `AssetsBalance`s.
+	 *
+	 * @param hash `BlockHash` to make call at
+	 * @param address `AccountId` associated with the balances
+	 * @param assets An array of `assetId`'s to be queried. If the length is zero
+	 * all assetId's associated to the account will be queried
+	 */
+	async fetchAssetBalances(
+		hash: BlockHash,
+		address: string,
+		assets: number[]
+	): Promise<IAccountAssetsBalances> {
+		const { api } = this;
+
+		const { number } = await api.rpc.chain.getHeader(hash);
+
+		let response;
+		if (assets.length === 0) {
+			/**
+			 * This will query all assets and return them in an array
+			 */
+			const keys = await api.query.assets.account.keysAt(hash);
+			const assetIds = this.extractAssetIds(keys);
+
+			response = await this.queryAssets(api, assetIds, address);
+		} else {
+			/**
+			 * This will query all assets by the requested AssetIds
+			 */
+			response = await this.queryAssets(api, assets, address);
+		}
+
+		const at = {
+			hash,
+			height: number.unwrap().toString(10),
+		};
+
+		return {
+			at,
+			assets: response,
+		};
+	}
+
+	/**
+	 * Fetch all `AccountApproval`'s with a given `AssetId` and a `AssetApprovalKey`
+	 * which consists of a `delegate` and an `owner(AccoundId)`
+	 *
+	 * @param hash `BlockHash` to make call at
+	 * @param address `AccountId` or owner associated with the approvals
+	 * @param assetId `AssetId` associated with the `AssetApproval`
+	 * @param delegate `delegate`
+	 */
+	async fetchAssetApproval(
+		hash: BlockHash,
+		address: string,
+		assetId: number,
+		delegate: string
+	): Promise<IAccountAssetApproval> {
+		const { api } = this;
+
+		/**
+		 * AssetApprovalKey, contains the `accountId` as the address key, and the
+		 * delegate `accountId`. Both are required for a successfull query
+		 */
+		const approvalKey = { owner: address, delegate };
+
+		const [{ number }, assetApproval] = await Promise.all([
+			api.rpc.chain.getHeader(hash),
+			api.query.assets.approvals(assetId, approvalKey),
+		]);
+
+		const { amount, deposit } = assetApproval.unwrap();
+
+		const at = {
+			hash,
+			height: number.unwrap().toString(10),
+		};
+
+		return {
+			at,
+			amount,
+			deposit,
+		};
+	}
+
+	/**
+	 * Takes in an array of AssetId's, and an AccountId and returns
+	 * all balances tied to those AssetId's.
+	 *
+	 * @param api ApiPromise
+	 * @param assets An Array of `AssetId`'s or numbers representing assetId's
+	 * @param address An `AccountId` associated with the queried path
+	 */
+	async queryAssets(
+		api: ApiPromise,
+		assets: AssetId[] | number[],
+		address: string
+	): Promise<IAssetBalance[]> {
+		const queryInputAssets = async (): Promise<AssetBalance[]> =>
+			await Promise.all(
+				assets.map((assetId: number | AnyNumber) =>
+					api.query.assets.account(assetId, address)
+				)
+			);
+
+		const response = (await queryInputAssets()).map((asset, i) => {
+			return {
```

Comment:
Why do we need to reformat `AssetBalance` like this? Should we just return the struct as is?

Reply:
If we return the `Struct` as is there is no AssetId attached to it.

Reply:
Makes sense, then for `extra` we can do something like `extra: asset.extra ?? undefined`, but need to figure out the appropriate way to check what `extra` is

Reply:
I would prefer if we treated the pallet as generically as possible and not make too many assumptions about the runtime it lives in.